src/plugins/vis_types/xy/public/expression_functions/x_domain.ts

```diff
@@ -0,0 +1,181 @@
+  const minInterval = getAdjustedInterval(
+    values,
+    domain.intervalValue,
+    domain.intervalUnit as unitOfTime.Base,
+    timezone ?? ''
+  );
+
+  return {
+    min: domainMin,
+    max: domainMax,
+    minInterval,
+  };
+};
+
+export const xDomain = (): ExpressionFunctionDefinition<
+  'x_domain',
+  Datatable | null,
+  XDomainArguments,
+  ExpressionValueXDomain
+> => ({
+  name: X_DOMAIN_EXPRESSION,
+  help: i18n.translate('visTypeXy.function.valueaxis.help', {
+    defaultMessage: 'Generates value axis object',
+  }),
+  type: X_DOMAIN_EXPRESSION,
+  args: {
+    minInterval: {
+      types: ['number'],
+      help: i18n.translate('visTypeXy.function.xDomain.minInterval.help', {
+        defaultMessage: 'Min interval',
+      }),
+    },
+
+    intervalUnit: {
+      types: ['string'],
+      help: i18n.translate('visTypeXy.function.xDomain.intervalUnit.help', {
+        defaultMessage: 'intervalUnit',
+      }),
+    },
+
+    intervalValue: {
+      types: ['number'],
+      help: i18n.translate('visTypeXy.function.xDomain.intervalValue.help', {
+        defaultMessage: 'intervalValue',
+      }),
+    },
+
+    column: {
+      types: ['vis_dimension'],
+      help: i18n.translate('visTypeXy.function.column.intervalValue.help', {
+        defaultMessage: 'column',
+      }),
+    },
+
+    timezone: {
+      types: ['string'],
+      help: i18n.translate('visTypeXy.function.xDomain.timezone.help', {
+        defaultMessage: 'timezone',
+      }),
+    },
+
+    considerInterval: {
+      types: ['boolean'],
+      help: i18n.translate('visTypeXy.function.xDomain.considerInterval.help', {
+        defaultMessage: 'considerInterval',
+      }),
+    },
+
+    min: {
+      types: ['number'],
+      help: i18n.translate('visTypeXy.function.xDomain.min.help', {
+        defaultMessage: 'min',
+      }),
+    },
+
+    max: {
+      types: ['number'],
+      help: i18n.translate('visTypeXy.function.xDomain.max.help', {
+        defaultMessage: 'max',
+      }),
+    },
+
+    logBase: {
+      types: ['string'],
+      help: i18n.translate('visTypeXy.function.xDomain.logBase.help', {
+        defaultMessage: 'logBase',
+      }),
+    },
+
+    coordinates: {
+      types: ['string', 'number'],
+      help: i18n.translate('visTypeXy.function.xDomain.coordinates.help', {
+        defaultMessage: 'coordinates',
+      }),
+      multi: true,
+    },
+  },
+  fn: (context, args) => {
+    const {
+      min,
+      max,
+      minInterval,
+      logBase,
+      coordinates,
+      column,
+      intervalUnit,
+      intervalValue,
+      timezone,
+      considerInterval,
+    } = args;
+
+    const domain = { min, max, minInterval, logBase, coordinates };
+
+    if (
+      intervalUnit &&
+      intervalValue !== undefined &&
+      min !== undefined &&
+      max !== undefined &&
+      column !== undefined
+    ) {
+      const adjusted = getAdjustedDomain(
+        context?.rows ?? [],
```

Comment:
getAdjustedDomain needs the actual data table to return the right result. Previously this was simple because it would be part of the renderer, but as you moved it into an expression function, you need to get access to the data table in a different way. In the declaration of x_domain you simply made the datatable input optional, but in practice (in the to_ast file) it will never receive an input so it can’t do its calculations based on the data table - this is the line where you fall back to an empty array instead of the actual “rows” which always will happen. We need to find a solution for this general problem (which will only get worse as in the Lens use case a lot of things are decided based on the actual data to render instead of static arguments). I would like to do this in a proper design document rather than a large, hard to review PR

Reply:
About this issue, possibly, I'm wrong, but the function, which is going after some `datasource` expression function (as `es_aggs`, `es_docs`, etc.) is accepting the actual data, which came from it and spreading to all arguments, which are expressions and spreading came data to the next expression in the pipe of the expressions and it is going forward without change. If I'm wrong, I'd like to be fixed by Peter )

Reply:
And fetching new data is executing new cycle of executing `datasource` expression, passing the data to the other expressions in the pipe and updating the renderer, so, we'll have the updated data all the time ( at list, at vis_editors and canvas, I don't really know, if Lens is using it in some specific way).

Reply:
Ah, you are right, of course, thanks for explaining - shouldn’t have reviewed so late in the evening :)

Reply:
what is the reason for moving this from renderer into the expression function ? I would prefer to keep rendering logic in the actual renderer and keep the expression function simple (just providing configuration to the renderer)

Reply:
yea, I would like to do that too, but adjustedDomain is a specific logic, related to the aggregations, to be supported for vis_types. I wished to keep the renderer clean from any logic, related to the aggregations.